src/cloud/apis/reporting.ts

```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
     }
   }
 }
+
+const asRudderOptions = (tags: PointTags, campaign: object) => {
+  const options = {}
+  let context = null
+  Object.keys(tags).forEach((key: string) => {
+    const value = tags[key]
+    if (key != 'context') {
+      options[key] = value
+    } else {
+      if (typeof value != 'object') {
+        // context is reserved rudder element key and must be object, see rudder-sdk-js (#20463)
+        options['appContext'] = value
+      } else {
+        context = options['context'] = value
+      }
+    }
+  })
+  if (Object.keys(campaign).length) {
+    if (context == null) {
+      context = options['context'] = {}
+    }
+    context['campaign'] = campaign
+  }
+  return options
+}
```

Comment:
overall I'd love it if we can make this function a bit clearer. I understand that we're building out a rudderStackOption, but what does that look like? Is there a type we can reference that rudderstack's SDK provides? There are lots of nested conditionals that make this a little unclear what's happening where. Maybe that's a necessary evil, but maybe it doesn't have to be? We're also converting an object into an array in order to iterate over it, but why not just us a `for in` loop to iterate over the original object and deal with the data that way?

Reply:
I have refactored the whole function, hopefully it is better now. Unfortunately there is no reference in SDK that I could find. I had to use rudder SDK js source code and also their support on Slack channel(https://app.slack.com/client/T01E4PJL5J7/C01E4PLB135/thread/C01E4PLB135-1628077828.109800) 
![image](https://user-images.githubusercontent.com/42931850/129576458-82a30fa6-d2b6-4fe6-afa8-6c4d8dcf78a4.png)